tests/kernel/sched/schedule_api/src/test_sched_pending.c

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2021 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <ksched.h>
+#include <spinlock.h>
+#include <wait_q.h>
+#include <ztest.h>
+
+struct sched_pending {
+	_wait_q_t waitq;
+	struct k_spinlock lock;
+} pending;
+
+#define STACK_SIZE	(1024 + CONFIG_TEST_EXTRA_STACKSIZE)
+
+K_THREAD_STACK_DEFINE(test_stack, STACK_SIZE);
+
+static struct k_thread tdata;
+
+static atomic_t num;
+static k_tid_t tid;
+
+static void thread_handler1(void *p1, void *p2, void *p3)
+{
+	z_waitq_init(&pending.waitq);
+	k_spinlock_key_t key = k_spin_lock(&pending.lock);
+
+	z_pend_thread(_current, &pending.waitq, K_FOREVER);
+	z_reschedule(&pending.lock, key);
+}
+
+static void thread_handler2(void *p1, void *p2, void *p3)
+{
+	z_waitq_init(&pending.waitq);
+
+	unsigned int key = irq_lock();
+
+	z_pend_curr_irqlock(key, &pending.waitq, K_FOREVER);
+}
+
+static void thread_handler3(void *p1, void *p2, void *p3)
+{
+	atomic_inc(&num);
+}
+
+/**
+ * @brief Test kernel api z_pend_thread()  z_unpend_thread()
+ * z_unpend_all() z_pend_curr_irqlock()
+ *
+ * @details Part1: Creat a child thread and when it's running,
+ * use z_pend_thread() api to set it to pending state.
+ * Verfiy the child thread state in main thread, it should be in pending.
+ * Use z_unpend_thread() api to unpend the child thread and verify in
+ * main thread. It should be in unpending state.
+ * Part2: Creat a child thread and make it in pending state.
+ * Use z_unpend_all() to unpend it then verify the state.
+ * Part3: Test z_pend_curr_irqlock() and verify it's pending state.
+ *
+ * @ingroup kernel_sched_tests
+ *
+ * @see z_pend_thread() z_unpend_thread() z_unpend_all()
+ * z_pend_curr_irqlock()
+ */
+void test_kernel_api_pend_unpend(void)
+{
+	/* Part1: test kernel api z_pend_thread() z_unpend_thread() */
+	tid = k_thread_create(&tdata, test_stack, STACK_SIZE,
+					thread_handler1, NULL, NULL, NULL,
+					K_PRIO_PREEMPT(0), 0, K_NO_WAIT);
+	/* relinquish CPU for above thread to start */
+	k_msleep(1);
+	zassert_true(z_is_thread_pending(&tdata), NULL);
+	z_unpend_thread(tid);
+	z_ready_thread(tid);
+	zassert_false(z_is_thread_pending(&tdata), NULL);
```

Comment:
Here's a good example: What coverage problem are you trying to address here?  All the calls here should be exercised by straightforward k_sem_{take,give}() or any other IPC primitive.  We absolutely have coverage for pend/unpend/ready, there's no way to do IPC without it.  Is it just the entry points you're trying to hit?

Reply:
The background of why I write these testcode because we are doing module test separately. If we run "twister -p qemu_x86 -T tests/kernel/sched/schedule_api. We can found there are many lines not coveraged. Even this line is covered when run "twister -p qemu_x86 -T tests/kernel". Our coverage range only depend on the moudles. That's our standard of doing moule tests. 
As you mentioned, it is not appropriate to call these internal APIs directly.  I'll delete these code which directly test z_prefix  
function.  Thank you for your comments.

Reply:
But please submit a bug calling out the missing coverage in sched.c.  It's probably dead code that should be removed.

Reply:
I'm not sure when a internal api only used in testcase but it has never been used in kernel is a piece of dead code. For example "z_move_thread_to_end_of_prio_q()".  What do you think?

Reply:
If it's not showing up in coverage we should find out why.  Please file the bug :)

Reply:
HI, @andyross . The function "z_move_thread_to_end_of_prio_q" I mentioned above is useful I think because it is called in arm_isr_handler().  After Yerabolu's PR #32891 has been merged. At present, there are no dead code in kernel/sched.c.
I have remove those code which directly test z_prefix function. The remaining code is about negative case in sched.c, could you help to review it. Thanks.